# Building Question Answering Bot using RAG and PineCONE VectorDB




---



### Use Wikipedia API to fetch relevant documents and articles related to business and creating pdf which in future use for creating VectorDB
----

In [ ]:
pip install wikipedia-api

In [22]:
import wikipediaapi

# Define a user agent
user_agent = "MyWikipediaBot/1.0"

# Set the headers with the user agent
headers = {
    'User-Agent': user_agent
}

# Initialize the Wikipedia API with the headers
wiki_wiki = wikipediaapi.Wikipedia('en', headers=headers)

# Now start Fetching a Wikipedia page


In [23]:
# list of all relevant keyword for extracting business docs from wikipedia api
business_keywords = [
    "Business model", "Competitive strategy", "Strategic planning", "SWOT analysis", "Business growth strategy",
    "Market penetration", "Value proposition", "Key performance indicators", "Corporate governance",
    "Business operations", "Operational efficiency", "Mission statement", "Vision statement", "Management style",
    "Leadership", "Strategic alliances", "Business expansion", "Business restructuring", "Business transformation",
    "Market positioning", "Risk management", "Scenario planning", "Portfolio management", "Disruptive innovation",
    "Value chain", "Lean management", "Business sustainability", "Corporate social responsibility", "customer support",

    "Brand management", "Brand equity", "Target market", "Customer segmentation", "Consumer behavior", "Marketing strategy",
    "Content marketing", "Influencer marketing", "Digital marketing", "Marketing automation", "Customer loyalty programs",
    "Customer relationship management", "Social media marketing", "SEO", "SEM", "Affiliate marketing", "Product lifecycle management",
    "Viral marketing", "Email marketing", "Conversion rate optimization", "Lead generation", "Marketing campaign",
    "Public relations", "Advertising", "Customer experience", "Omnichannel marketing", "Market research", "User-generated content",

    "Sales funnel", "Sales strategy", "Lead nurturing", "Sales pipeline", "Closing deals", "Cold calling",
    "Business development", "Sales negotiation", "Prospecting", "Sales targets", "Customer acquisition cost",
    "Sales forecasting", "Upselling and cross-selling", "B2B sales", "B2C sales", "Enterprise sales",
    "Account management", "Key account management", "Channel sales", "Sales territory management", "Sales enablement",
    "Sales performance", "Product demos", "Sales presentations", "Customer referrals",

    "Financial analysis", "Income statement", "Balance sheet", "Cash flow statement", "Profit and loss", "Budgeting",
    "Capital budgeting", "Taxation", "Financial reporting", "Audit", "Payroll management", "Cost accounting",
    "Financial forecasting", "Business valuation", "Investment portfolio", "Risk-adjusted return", "ROI",
    "Working capital management", "Revenue generation", "Profit margin", "Cash management", "Debt management",
    "Corporate tax", "Accounting principles", "Cost-benefit analysis", "Financial modeling", "Business credit",
    "Mergers and acquisitions",

    "Recruitment", "Talent acquisition", "Employee onboarding", "Job descriptions", "Employee benefits", "Employee training",
    "Performance appraisal", "Compensation and benefits", "HR analytics", "Employee retention", "Employee engagement",
    "Labor laws", "Workplace diversity", "Employee wellness programs", "Organizational culture", "Employee motivation",
    "Leadership development", "Succession planning", "Team management", "HR policies", "Workforce planning", "HR technology",
    "Labor market", "Employee relations",

    "Helpdesk support", "Customer service", "Customer satisfaction", "Support ticketing system", "Knowledge base",
    "FAQ", "Live chat", "Self-service portal", "Troubleshooting", "Call center", "Customer feedback", "Customer retention",
    "Service level agreement", "Service quality", "Warranty", "Refund policies", "Returns management", "Customer complaints",
    "Escalation process", "Customer loyalty programs", "User manual", "Product documentation", "Feedback loop", "Customer success management",

    "Cloud computing", "AI", "Machine learning", "Data science", "Big data", "Internet of Things", "Blockchain technology",
    "Augmented Reality", "Virtual Reality", "Cybersecurity", "Data privacy", "Digital transformation", "Smart contracts",
    "Software as a Service", "Business automation", "IT infrastructure", "Business intelligence", "Cloud storage", "Data analytics",
    "Cyber resilience", "IT support", "API integration", "Network security", "IT governance",

    "Business law", "Contract law", "Intellectual property", "Non-disclosure agreement", "Non-compete agreement",
    "Business incorporation", "Patent law", "Trademark law", "Data protection", "General Data Protection Regulation",
    "Compliance audits", "Labor law", "Employment contracts", "Legal compliance", "Business insurance", "Regulatory reporting",
    "Risk management", "Licensing and permits", "Corporate ethics", "Anti-money laundering", "Anti-bribery and corruption",

    "E-commerce platform", "Online store", "Payment gateway", "Shopping cart", "Digital wallet", "Product catalog",
    "Inventory management", "Shipping and logistics", "Dropshipping", "Payment processing", "Product reviews and ratings",
    "Affiliate marketing", "Price optimization", "Online advertising", "Product recommendations", "Retail strategy",
    "Omnichannel retailing", "Consumer electronics", "Wholesale vs retail", "Supply chain management", "Product bundling",

    "Supply chain management", "Logistics", "Inventory control", "Procurement", "Product sourcing", "Supplier relationship management",
    "Distribution channels", "Warehouse management", "Stock management", "Order fulfillment", "Demand forecasting",
    "Supplier negotiation", "Lean operations", "Just-in-time", "Production planning", "Shipping and delivery", "Product packaging",
    "Supply chain optimization", "Freight management",

    "Startup incubator", "Entrepreneur", "Angel investors", "Seed funding", "Venture capital", "Business pitch", "Bootstrapping",
    "Startup ecosystem", "Lean startup", "Crowdfunding", "Entrepreneurial mindset", "Startup accelerator", "Business angel",
    "Startups and innovation", "Founder", "Co-founder", "Business incubators", "Startup valuation",

    "Profit", "Revenue", "Market share", "Business growth", "Business model innovation", "Return on equity", "Cash flow management",
    "Business restructuring", "Mergers and acquisitions", "Customer segmentation", "Branding", "Product differentiation",
    "Competitive advantage", "Exit strategy", "Market research", "Customer loyalty", "Business ethics","Business",
    "Marketing",
    "Finance",
    "Entrepreneurship",
    "Economics",
    "Corporate_governance",
    "Supply_chain_management",
    "Financial_statements",
    "Investment",
    "Business_model"
]


In [24]:

# Initialize a string to store all summaries
all_summaries = ""
cnt=0
c=0
# Iterate through the keywords and fetch their summaries
for keyword in business_keywords:
    cnt+=1
    page = wiki_wiki.page(keyword)
    if page.exists():
        summary = page.summary
        all_summaries += f"Title: {page.title}\nSummary: {summary}\nURL: {page.fullurl}\n{'='*100}\n"
    else:
        c+=1
        print(f"Page '{keyword}' does not exist.")

# Now,all_summaries will contain the combined summaries for all the pages.
# save it to a file for further analysis or storage
with open('business_summary_wikipedia_data.txt', 'w') as f:
    f.write(all_summaries)

cnt-c

Page 'Business growth strategy' does not exist.
Page 'Business operations' does not exist.
Page 'Business expansion' does not exist.
Page 'Business restructuring' does not exist.
Page 'Business sustainability' does not exist.
Page 'Customer loyalty programs' does not exist.
Page 'Omnichannel marketing' does not exist.
Page 'Sales strategy' does not exist.
Page 'Closing deals' does not exist.
Page 'Sales negotiation' does not exist.
Page 'Sales targets' does not exist.
Page 'Upselling and cross-selling' does not exist.
Page 'B2B sales' does not exist.
Page 'B2C sales' does not exist.
Page 'Account management' does not exist.
Page 'Key account management' does not exist.
Page 'Channel sales' does not exist.
Page 'Sales territory management' does not exist.
Page 'Sales enablement' does not exist.
Page 'Sales performance' does not exist.
Page 'Product demos' does not exist.
Page 'Sales presentations' does not exist.
Page 'Customer referrals' does not exist.
Page 'Payroll management' does n

222

In [25]:
cnt

288

In [26]:
c

66

In [27]:
with open('business_summary_wikipedia_data.txt', 'r') as file:
    for line in file:
        print(line.strip())  # .strip() use to remove leading/trailing whitespace if present


Title: Business model
Summary: A business model describes how a business organization creates, delivers, and captures value, in economic, social, cultural or other contexts. The model describes the specific way in which the business conducts itself, spends, and earns money in a way that generates profit. The process of business model construction and modification is also called business model innovation and forms a part of business strategy.
In theory and practice, the term business model is used for a broad range of informal and formal descriptions to represent core aspects of an organization or business, including purpose, business process, target customers, offerings, strategies, infrastructure, organizational structures, profit structures, sourcing, trading practices, and operational processes and policies including culture.
URL: https://en.wikipedia.org/wiki/Business_model
Title: Strategic management
Summary: In the field of management, strategic management involves the formulatio

In [28]:

# Initialize a string to store full text not just summary
all_text = ""
cnt=0
c=0
# Iterate through the keywords and fetch all text one by one for each keyword
for keyword in business_keywords:
    cnt+=1
    page = wiki_wiki.page(keyword)
    if page.exists():
        text = page.text  # Get the entire text
        all_text += f"Title: {page.title}\nFull text: {text}\nURL: {page.fullurl}\n{'='*100}\n"
    else:
        c+=1
        print(f"Page '{keyword}' does not exist.")

# Now, `all_text` will contain the combined text data for all the keywords
# save it to a file for further analysis or storage
with open('business_full_text_wikipedia_data.txt', 'w') as f:
    f.write(all_text)

cnt-c

Page 'Business growth strategy' does not exist.
Page 'Business operations' does not exist.
Page 'Business expansion' does not exist.
Page 'Business restructuring' does not exist.
Page 'Business sustainability' does not exist.
Page 'Customer loyalty programs' does not exist.
Page 'Omnichannel marketing' does not exist.
Page 'Sales strategy' does not exist.
Page 'Closing deals' does not exist.
Page 'Sales negotiation' does not exist.
Page 'Sales targets' does not exist.
Page 'Upselling and cross-selling' does not exist.
Page 'B2B sales' does not exist.
Page 'B2C sales' does not exist.
Page 'Account management' does not exist.
Page 'Key account management' does not exist.
Page 'Channel sales' does not exist.
Page 'Sales territory management' does not exist.
Page 'Sales enablement' does not exist.
Page 'Sales performance' does not exist.
Page 'Product demos' does not exist.
Page 'Sales presentations' does not exist.
Page 'Customer referrals' does not exist.
Page 'Payroll management' does n

222

In [29]:
with open('business_full_text_wikipedia_data.txt', 'r') as file:
    for line in file:
        print(line.strip())


Streaming output truncated to the last 5000 lines.
Elements
Entrepreneurship includes the creation or extraction of economic value. It is the act of being an entrepreneur, or the owner or manager of a business enterprise who, by risk and initiative, attempts to make profits. Entrepreneurs act as managers and oversee the launch and growth of an enterprise. Entrepreneurship is the process by which either an individual or a team identifies a business opportunity and acquires and deploys the necessary resources required for its exploitation.
In the early 19th century, the French economist Jean-Baptiste Say provided a broad definition of entrepreneurship, saying that it "shifts economic resources out of an area of lower and into an area of higher productivity and greater yield". Entrepreneurs create something new and unique—they change or transmute value.
Regardless of the firm size, big or small, it can take part in entrepreneurship opportunities. There are four criteria for becoming an en

In [31]:
# Use to convert .txt into .pdf
!pip install fpdf

  Preparing metadata (setup.py) ... done
  Created wheel for fpdf: filename=fpdf-1.7.2-py2.py3-none-any.whl size=40704 sha256=f9a2a25be45552635f95d3106d47059b661bfcad8b0a36952a00934b6a728e9d
  Stored in directory: /root/.cache/pip/wheels/f9/95/ba/f418094659025eb9611f17cbcaf2334236bf39a0c3453ea455
Successfully built fpdf


In [32]:
from fpdf import FPDF


def txt_to_pdf(txt_file, pdf_file):
    # Initialize FPDF object
    pdf = FPDF()
    pdf.set_auto_page_break(auto=True, margin=15)
    pdf.add_page()

    # Set font for the PDF (Arial, size 12)
    pdf.set_font("Arial", size=12)

    # Set margins and line height for better readability
    pdf.set_left_margin(10)
    pdf.set_right_margin(10)
    line_height = 10


    with open(txt_file, "r") as file:

        lines = file.readlines()  # Read the content of the .txt file

        # Write each line of the text file into the PDF
        for line in lines:
            # Add line to the PDF with text wrapping
            pdf.multi_cell(0, line_height, line.encode('latin-1', 'replace').decode('latin-1'))

    # Save the PDF
    pdf.output(pdf_file)
    print(f"PDF saved as {pdf_file}")



#convert .txt into .pdf file
txt_to_pdf("business_full_text_wikipedia_data.txt", "business_full_text_wikipedia.pdf")


PDF saved as business_full_text_wikipedia.pdf


In [33]:
txt_to_pdf("business_summary_wikipedia_data.txt", "business_summary_wikipedia.pdf")

PDF saved as business_summary_wikipedia.pdf


### Start Building QA Chatbot using LangChain
---

In [ ]:
!pip install langchain pinecone

In [35]:
import langchain
import pinecone

In [36]:
!pip install -U langchain-community

In [ ]:
from langchain.document_loaders import PyPDFDirectoryLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter
from langchain.vectorstores import Pinecone

In [ ]:
import os
!pip install pypdf

In [39]:
# step 1: convert text into ducment object and read document object
def read_doc(directory):
  loader=PyPDFDirectoryLoader(directory)
  documents=loader.load()
  return documents

In [40]:
document=read_doc('/content/')

In [41]:
len(document)

161

In [42]:
document

[Document(metadata={'source': '/content/business_summary_wikipedia.pdf', 'page': 0}, page_content='Title: Business model\nSummary: A business model describes how a business organization creates, delivers, and captures\nvalue, in economic, social, cultural or other contexts. The model describes the specific way in which\nthe business conducts itself, spends, and earns money in a way that generates profit. The process\nof business model construction and modification is also called business model innovation and forms\na part of business strategy.\nIn theory and practice, the term business model is used for a broad range of informal and formal\ndescriptions to represent core aspects of an organization or business, including purpose, business\nprocess, target customers, offerings, strategies, infrastructure, organizational structures, profit\nstructures, sourcing, trading practices, and operational processes and policies including culture.\nURL: https://en.wikipedia.org/wiki/Business_model\

In [43]:
## step 2: divide doc into chunks
def chunk_data(docs,chunk_size=500,chunk_overlap=40):
  text_splitter=RecursiveCharacterTextSplitter(chunk_size=chunk_size,chunk_overlap=chunk_overlap)
  chunks=text_splitter.split_documents(docs)
  return chunks

documents_chunk=chunk_data(docs=document)

In [44]:
documents_chunk

[Document(metadata={'source': '/content/business_summary_wikipedia.pdf', 'page': 0}, page_content='Title: Business model\nSummary: A business model describes how a business organization creates, delivers, and captures\nvalue, in economic, social, cultural or other contexts. The model describes the specific way in which\nthe business conducts itself, spends, and earns money in a way that generates profit. The process\nof business model construction and modification is also called business model innovation and forms\na part of business strategy.'),
 Document(metadata={'source': '/content/business_summary_wikipedia.pdf', 'page': 0}, page_content='a part of business strategy.\nIn theory and practice, the term business model is used for a broad range of informal and formal\ndescriptions to represent core aspects of an organization or business, including purpose, business\nprocess, target customers, offerings, strategies, infrastructure, organizational structures, profit\nstructures, sourcin

In [45]:
# document chunk which is created of size within 500
len(documents_chunk)

804

In [46]:
#all document object ie data file
len(document)

161

In [47]:
## step 3:  vector embedding for this chunk to convert into numerical data
from sentence_transformers import SentenceTransformer

# Initialize the SentenceTransformer model
model = SentenceTransformer('all-MiniLM-L6-v2')


/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_auth.py:94: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


modules.json:   0%|          | 0.00/349 [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

README.md:   0%|          | 0.00/10.7k [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

config.json:   0%|          | 0.00/612 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/90.9M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/350 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/232k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/466k [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/190 [00:00<?, ?B/s]

In [48]:
# extract plain text for doc chunk which is passed to sentence to for create embedding &
# embedding need plain text not document whic include plain text and plus meta data
text_chunks = [doc.page_content for doc in documents_chunk]


In [49]:
documents_chunk

[Document(metadata={'source': '/content/business_summary_wikipedia.pdf', 'page': 0}, page_content='Title: Business model\nSummary: A business model describes how a business organization creates, delivers, and captures\nvalue, in economic, social, cultural or other contexts. The model describes the specific way in which\nthe business conducts itself, spends, and earns money in a way that generates profit. The process\nof business model construction and modification is also called business model innovation and forms\na part of business strategy.'),
 Document(metadata={'source': '/content/business_summary_wikipedia.pdf', 'page': 0}, page_content='a part of business strategy.\nIn theory and practice, the term business model is used for a broad range of informal and formal\ndescriptions to represent core aspects of an organization or business, including purpose, business\nprocess, target customers, offerings, strategies, infrastructure, organizational structures, profit\nstructures, sourcin

In [47]:
text_chunks

['Title: Business model\nSummary: A business model describes how a business organization creates, delivers, and captures\nvalue, in economic, social, cultural or other contexts. The model describes the specific way in which\nthe business conducts itself, spends, and earns money in a way that generates profit. The process\nof business model construction and modification is also called business model innovation and forms\na part of business strategy.',
 'a part of business strategy.\nIn theory and practice, the term business model is used for a broad range of informal and formal\ndescriptions to represent core aspects of an organization or business, including purpose, business\nprocess, target customers, offerings, strategies, infrastructure, organizational structures, profit\nstructures, sourcing, trading practices, and operational processes and policies including culture.\nURL: https://en.wikipedia.org/wiki/Business_model',
 "============================================================

In [50]:
# embeddings function
def generate_embeddings_for_chunks(text_chunks):
    embeddings = model.encode(text_chunks)
    return embeddings

# Generate embeddings for each chunk
chunk_embeddings = generate_embeddings_for_chunks(text_chunks)

In [51]:
#text chunk ie only plain text for each doc chunks
len(text_chunks)

804

In [52]:
len(chunk_embeddings[0])

384

In [53]:
chunk_embeddings

array([[ 0.03715841, -0.05087739, -0.03846387, ..., -0.04792089,
         0.07053431,  0.00072295],
       [ 0.08693385, -0.0290452 , -0.01876907, ...,  0.01864725,
         0.0453697 ,  0.01582074],
       [ 0.03376321,  0.05630676, -0.04452125, ...,  0.02267691,
        -0.00169465,  0.0365942 ],
       ...,
       [ 0.06423485, -0.04352978, -0.05286697, ..., -0.04603238,
         0.0419427 ,  0.01739166],
       [ 0.04042291, -0.03562506, -0.03961556, ..., -0.0171605 ,
         0.10702675,  0.02498419],
       [ 0.08328034, -0.07886633, -0.07130105, ...,  0.00560425,
         0.04991134,  0.06100166]], dtype=float32)

#Use 384 as dimension un pinecone because vector create of dimension 384 by sentence transformer

In [52]:
# step 4: create Vector Search DB in Pinecone

In [54]:
import os
from pinecone import Pinecone

# Initialize Pinecone client using your API key
pc = Pinecone(api_key="****************************************")

# Connect to your existing index or create index if not exist
index_name = "********************************"
index = pc.Index(index_name)


In [55]:
index

In [ ]:
# Now you can interact with Pinecone (e.g., list indexes)
indexes = pc.list_indexes()
print(indexes)

In [57]:
len(chunk_embeddings)

804

In [244]:

# Each metadata entry contains the text for that chunk
# 'chunk_embeddings' contains vector embeddings and 'text_chunks' contains the text for metadata
# Create sequential IDs
ids = [f"chunk_{i}" for i in range(len(chunk_embeddings))]

# Create a dictionary for metadata (store the original text)
metadata = [{"text": text} for text in text_chunks]

# Upsert embeddings along with metadata
index.upsert(vectors=zip(ids, chunk_embeddings, metadata))




{'upserted_count': 805}

In [ ]:
ids

In [246]:
len(ids)

805

In [247]:
len(chunk_embeddings)

805

In [58]:
# Check the status of the index after upsert
result = index.describe_index_stats()
print(result)


{'dimension': 384,
 'index_fullness': 0.0,
 'namespaces': {'': {'vector_count': 805}},
 'total_vector_count': 805}


In [59]:
# Cosine similary retrive results ie retrive data related to query from vectordb
def retrive_query(query,k=3):
  query_embedding=model.encode([query])[0]
  query_embedding = query_embedding.tolist()
  results=index.query(vector=query_embedding,top_k=k,include_metadata=True)
  return results

In [60]:

# Example query
query = "What is Supply Chain Management?"
res=retrive_query(query)
res


{'matches': [{'id': 'chunk_624',
              'metadata': {'text': 'customers. A more narrow definition of '
                                   'supply chain management is the "design, '
                                   'planning,\n'
                                   'execution, control, and monitoring of '
                                   'supply chain activities with the objective '
                                   'of creating net value,\n'
                                   'building a competitive infrastructure, '
                                   'leveraging worldwide logistics, '
                                   'synchronising supply with\n'
                                   'demand and measuring performance '
                                   'globally". This can include the movement '
                                   'and storage of raw\n'
                                   'materials, work-in-process inventory, '
                                   'finished goo

In [66]:
!pip install huggingface

In [59]:
from langchain.chains import LLMChain
from langchain.llms import HuggingFacePipeline
from transformers import pipeline
from langchain.prompts import PromptTemplate
from langchain.schema import Document
from sentence_transformers import SentenceTransformer

In [60]:
"""
t5-base
t5-large
facebook/bart-large
google/pegasus-xsum

google/pegasus-large

allenai/longformer-large-4096
"""

# explore all these but google/pegasus-xsum give better result among all !

'\nt5-base\nt5-large\nfacebook/bart-large\ngoogle/pegasus-xsum\n\ngoogle/pegasus-large\n\nallenai/longformer-large-4096\n'

In [61]:
# Function to initialize the QA model pipeline
def initialize_qa_pipeline(model_name="google/pegasus-xsum"):
    return pipeline("text2text-generation", model=model_name)

# Initialize the QA model pipeline by default model used is google/pegasus-xsum
# but you can change this model with any other model
qa_pipeline = initialize_qa_pipeline("google/pegasus-xsum")

config.json:   0%|          | 0.00/1.39k [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/2.28G [00:00<?, ?B/s]

Some weights of PegasusForConditionalGeneration were not initialized from the model checkpoint at google/pegasus-xsum and are newly initialized: ['model.decoder.embed_positions.weight', 'model.encoder.embed_positions.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


generation_config.json:   0%|          | 0.00/259 [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/87.0 [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/1.91M [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/3.52M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/65.0 [00:00<?, ?B/s]

Device set to use cpu


In [65]:

# Wrap the Hugging Face pipeline with HuggingFacePipeline from Langchain

llm = HuggingFacePipeline(pipeline=qa_pipeline)

# design prompt for effective and accurate response
prompt_template = """
You are an intelligent and trustworthy assistant. Given the provided context, answer the question clearly, concisely, and in grammatically correct sentences.
If the context does not provide sufficient information, use your knowledge to add relevant and accurate details.
However, if you genuinely do not know the answer, state that honestly instead of making up information.
Use the subject-verb-object format in all sentences and ensure proper punctuation.

NOTE: IF YOU ARE NOT 100% SURE OF THE ANSWER, SAY, "I am really sorry but I don't have enough information to answer this question." DO NOT GUESS OR FABRICATE DETAILS UNDER ANY CIRCUMSTANCES.
Make sure the response:

1. Is written in proper English with correct grammar and punctuation.
2. Adheres to the facts and concepts provided in the context.
3. Includes additional knowledge when the context is insufficient (if available).
4. Avoids assumptions or fabricated details.
5. Uses the subject-verb-object format in all sentences and ensures proper punctuation.
6. Access context by given URL if you needed additional information
Example:
Context: {context}

Question: {question}

Answer:
"""

# Creating a PromptTemplate with the defined structure
prompt = PromptTemplate(input_variables=["context", "question"], template=prompt_template)

# Initialize the LLMChain with the prompt and the model -llm
llm_chain = LLMChain(llm=llm, prompt=prompt)

# Function to retrieve relevant documents
# Cosine similary retrive results ie retrive data related to query from vectordb
def retrieve_query(query, k=3):

    # Initialize the model for embeddings (assumed to be Sentence-Transformer or similar)
    model = SentenceTransformer('all-MiniLM-L6-v2')
    query_embedding = model.encode([query])[0]
    query_embedding = query_embedding.tolist()

    # retrieve top k result based on query and store in results
    results = index.query(vector=query_embedding, top_k=k, include_metadata=True)
    return [match['metadata']['text'] for match in results['matches']]



# Function to retrieve and answer a query using LLMChain
def retrieve_and_answer(query):
    # Retrieve relevant documents
    retrieved_contexts = retrieve_query(query)

    # Check if any documents were retrieved
    if not retrieved_contexts:
        return "No relevant documents found."

    # Combine the retrieved documents into one string
    combined_context = " ".join([context.capitalize() + '.' for context in retrieved_contexts])

    # Use the QA pipeline to answer the query
    response = qa_pipeline(combined_context, query)

    response=response[0]['generated_text']

    if len(response.split()) < 5 :  # No context but lengthy response
        return "I am not confident enough to provide an accurate answer."
    if response:
     response = response.capitalize().strip()
     if not response.endswith('.'):
         response += '.'

    return response


###Now asked question and response got from above model

---

In [70]:
# Question 1
query = "What is Supply Chain Management?"
ans=retrieve_and_answer(query)
print("\nAnswer:\n", ans)

Ignoring args : ('What is Supply Chain Management?',)



Answer:
 A supply chain is a network of interconnected points in which goods and services are produced and delivered from one point in time to another.


In [76]:
# Question 2
query = "What is entrepreneur?"
ans=retrieve_and_answer(query)
print("\nAnswer:\n", ans)

Ignoring args : ('What is entrepreneur?',)



Answer:
 An entrepreneur is an individual who creates and/or invests in one or more businesses, bearing most of the risks and enjoying most of the rewards.


In [77]:
# Question 3
query = "What to do if I have 100 rupee to invest?"
ans=retrieve_and_answer(query)
print("\nAnswer:\n", ans)

Ignoring args : ('What to do if I have 100 rupee to invest?',)



Answer:
 A glossary of terms for the investment industry:.


In [79]:
# Question 4
query = "What is SWOT analysis?"
ans=retrieve_and_answer(query)
print("\nAnswer:\n", ans)

Ignoring args : ('What is SWOT analysis?',)



Answer:
 Swot analysis is a decision-making technique that identifies the strengths, weaknesses, opportunities, and threats of an organization or project.
